src/record/api.rs

```diff
@@ -55,7 +56,113 @@ pub enum Row {
   Map(Vec<(Row, Row)>) // List of key-value pairs
 }
 
+// Macro to generate type-safe get_xxx methods e.g. get_bool, get_short
+macro_rules! row_primitive_accessor {
+  ($METHOD:ident, $VARIANT:ident, $TY:ty) => {
+    pub fn $METHOD(&self) -> Result<$TY, ParquetError> {
+        match *self {
+            Row::$VARIANT(v) => Ok(v),
+            _ => Err(ParquetError::General(format!("Cannot access {} as {}", self.get_type_name(), stringify!($VARIANT))))
+        }
+    }
+  }
+}
+
 impl Row {
+  /// Get the type name
+  fn get_type_name(&self) -> &'static str {
+    match *self {
+      Row::Null => "Null",
+      Row::Bool(_) => "Bool",
+      Row::Byte(_) => "Byte",
+      Row::Short(_) => "Short",
+      Row::Int(_) => "Int",
+      Row::Long(_) => "Long",
+      Row::Float(_) => "Float",
+      Row::Double(_) => "Double",
+      Row::Str(_) => "Str",
+      Row::Bytes(_) => "Bytes",
+      Row::Timestamp(_) => "Timestamp",
+      Row::Group(_) => "Group",
+      Row::List(_) => "List",
+      Row::Map(_) => "Map",
+    }
+  }
+
+  row_primitive_accessor!(get_bool, Bool, bool);
+  row_primitive_accessor!(get_byte, Byte, i8);
+  row_primitive_accessor!(get_short, Short, i16);
+  row_primitive_accessor!(get_int, Int, i32);
+  row_primitive_accessor!(get_long, Long, i64);
+  row_primitive_accessor!(get_float, Float, f32);
+  row_primitive_accessor!(get_double, Double, f64);
+  row_primitive_accessor!(get_timestamp, Timestamp, u64);
+
+  /// Type-safe accessor for Str type
+  pub fn get_string(&self) -> Result<&String, ParquetError> {
```

Comment:
One thing I'm not quite sure is whether it's better to just panic here, instead of returning `Result`. It seems pretty rare that the error case would happen, given that one always (I think?) will hold the schema of the row before calling these methods.

Reply:
I think panic is pretty extreme. I think the user should decide if they want to panic or not. In non-test code they just need to add a `?` at the end of the line, so it doesn't make much difference to code verbosity.